### PR TITLE
chore: allow override data pipeline instance

### PR DESCRIPTION
**Pull request by mrehan27**
part of [MBL-284](https://linear.app/customerio/issue/MBL-284/setup-testing-environment-for-data-pipelines-module)

### Changes

- Added common method to update Analytics configurations from module configuration so it can be reused by tests
- Allowed overriding DataPipelines instance for testing purposes only
- Reset static DI graph component

### CODE REVIEW

File: datapipelines/src/main/kotlin/io/customer/sdk/CustomerIOBuilder.kt [hunk 1]
```diff
@@ -59,6 +59,9 @@ class CustomerIOBuilder(
     // Configuration options required for migration from earlier versions
     private var migrationSiteId: String? = null
 
+    // Overrides the default DataPipelines module instance with custom implementation
+    private var overrideDataPipelineInstance: ((DataPipelinesModuleConfig) -> CustomerIO)? = null
```
Comment (Shahroz16) [inline datapipelines/src/main/kotlin/io/customer/sdk/CustomerIOBuilder.kt:63]: I think if we were to utilize delegate from DiGraph we won't have to override manually?

File: datapipelines/src/main/kotlin/io/customer/sdk/android/CustomerIO.kt [hunk 2]
```diff
@@ -64,21 +66,14 @@ class CustomerIO private constructor(
         }
     }
 
-    private val analytics: Analytics = Analytics(
+    internal val analytics: Analytics = overrideAnalytics ?: Analytics(
```
Comment (Shahroz16) [inline datapipelines/src/main/kotlin/io/customer/sdk/android/CustomerIO.kt:69]: same concept goes here as mentioned above

File: datapipelines/src/main/kotlin/io/customer/sdk/CustomerIOBuilder.kt (hunk 1 above)
Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/sdk/CustomerIOBuilder.kt:63]: Yes, this is a good suggestion. The only problem is, we need `DataPipelinesModuleConfig` to create test instance. We can think of options to solve this problem though, but I think using `DiGraph` can simplify a lot of code in this PR. I'll try this out and get back soon.

Comment (mrehan27) [reply datapipelines/src/main/kotlin/io/customer/sdk/CustomerIOBuilder.kt]: I added some updates to `SDKComponent` as we discussed on call earlier, let me know if there are any concerns.

File: datapipelines/src/test/java/io/customer/datapipelines/extensions/DiGraphExt.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import io.customer.sdk.core.di.SDKComponent
+
+/**
+ * Registers the analytics factory to the SDKComponent.
+ * The key should match identifier defined in [SDKComponent.analyticsFactory].
+ * The method is placed in test sources to avoid exposing it to release builds.
+ */
+fun SDKComponent.registerAnalyticsFactory(
+    factory: (moduleConfig: DataPipelinesModuleConfig) -> Analytics
+) = registerDependency(identifier = "analyticsFactory") { factory }
```
Comment (Shahroz16) [inline datapipelines/src/test/java/io/customer/datapipelines/extensions/DiGraphExt.kt:14]: lets make the `analyticsFactory` an enum? to make sure we don't misspell

**Review by Shahroz16 — APPROVED** [review]

File: datapipelines/src/test/java/io/customer/datapipelines/extensions/DiGraphExt.kt (hunk 3 above)
Comment (mrehan27) [reply datapipelines/src/test/java/io/customer/datapipelines/extensions/DiGraphExt.kt]: Updated 👍🏻

**Pull request merged**